File: src/main/resources/ui/utils/companySettings.ts
import { companySettingsService } from '../services/api';

export type CompanySettings = {
  id?: number;
  companyName: string;
  companyAddress?: string;
  companyPhone?: string;
  companyEmail?: string;
  invoiceTitle?: string;
  footerNote?: string;
  taglineMm?: string;
  logoBase64?: string;
  voucherConfigJson?: string;
  salePrefix?: string;
  saleDigits?: number;
  purchasePrefix?: string;
  purchaseDigits?: number;
  bookingPrefix?: string;
  bookingDigits?: number;
};

const DEFAULT_SETTINGS: CompanySettings = {
  companyName: 'SSPD IT Solution Center',
  companyAddress: 'No. 38/Kha, 56 Ward, Lhaw Kar Main Road, South Dagon, Yangon',
  companyPhone: '09-252425319',
  companyEmail: '',
  invoiceTitle: 'Sales Invoice',
  footerNote: 'Thank you for your business',
  taglineMm: 'ဝန်ဆောင်မှုဌာန',
  logoBase64: '',
  salePrefix: 'INV',
  saleDigits: 5,
  purchasePrefix: 'PUR',
  purchaseDigits: 5,
  bookingPrefix: 'BK',
  bookingDigits: 6,
};

let cached: CompanySettings | null = null;
let inflight: Promise<CompanySettings> | null = null;

export const buildCompanyContact = (s: CompanySettings) => {
  const parts: string[] = [];
  if (s.companyPhone) parts.push(`Phone: ${s.companyPhone}`);
  if (s.companyAddress) parts.push(`Address: ${s.companyAddress}`);
  if (s.companyEmail) parts.push(`Email: ${s.companyEmail}`);
  return parts.join(' | ');
};

export const getCompanySettings = async (force = false): Promise<CompanySettings> => {
  if (!force && cached) return cached;
  if (inflight) return inflight;
  inflight = (async () => {
    try {
      const res = await companySettingsService.getSettings();
      if (res.success && res.data) {
        cached = { ...DEFAULT_SETTINGS, ...res.data };
        return cached;
      }
    } catch {}
    cached = DEFAULT_SETTINGS;
    return cached;
  })().finally(() => { inflight = null; }) as Promise<CompanySettings>;
  return inflight;
};

export const setCompanySettingsCache = (s: CompanySettings) => {
  cached = { ...DEFAULT_SETTINGS, ...s };
  window.dispatchEvent(new CustomEvent('company-settings-updated', { detail: cached }));
};

export const clearCompanySettingsCache = () => {
  cached = null;
};

export const getCachedCompanySettings = (): CompanySettings => cached ?? DEFAULT_SETTINGS;

File: src/main/java/org/sspd/servicemgmt/companysettingoptions/model/CompanySettings.java
package org.sspd.servicemgmt.companysettingoptions.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Cache;
import org.hibernate.annotations.CacheConcurrencyStrategy;

@Entity
@Table(name = "company_settings")
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class CompanySettings {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(name = "company_name", nullable = false, length = 255)
    private String companyName = "SSPD IT Solution Center";

    @Column(name = "company_address", length = 1000)
    private String companyAddress;

    @Column(name = "company_phone", length = 100)
    private String companyPhone;

    @Column(name = "company_email", length = 150)
    private String companyEmail;

    @Column(name = "invoice_title", length = 255)
    private String invoiceTitle = "Sales Invoice";

    @Column(name = "footer_note", length = 500)
    private String footerNote = "Thank you for your business";

    @Column(name = "tagline_mm", length = 255)
    private String taglineMm;

    @Lob
    @Column(name = "logo_base64", columnDefinition = "LONGTEXT")
    private String logoBase64;

    @Lob
    @Column(name = "voucher_config_json", columnDefinition = "LONGTEXT")
    private String voucherConfigJson;

    @Builder.Default
    @Column(name = "setup_complete")
    private Boolean setupComplete = false;

    @Builder.Default
    @Column(name = "sale_prefix", length = 20)
    private String salePrefix = "INV";

    @Builder.Default
    @Column(name = "sale_digits")
    private Integer saleDigits = 5;

    @Builder.Default
    @Column(name = "purchase_prefix", length = 20)
    private String purchasePrefix = "PUR";

    @Builder.Default
    @Column(name = "purchase_digits")
    private Integer purchaseDigits = 5;

    @Builder.Default
    @Column(name = "booking_prefix", length = 20)
    private String bookingPrefix = "BK";

    @Builder.Default
    @Column(name = "booking_digits")
    private Integer bookingDigits = 6;
}
